drift/core/psmc.py

```diff
@@ -209,6 +213,273 @@ def _work_fisher_bias_m(self, mi):
+
+                # Sent data to (rank + 1) % size, hence local mi must be updated to (mi -1) % size
+                mi = low_bound[ci] + (mi - 1) % size
+
+            etallq = time.time()
+            print(
+                "Time needed for quadratic estimation on all ranks for this m-chunk",
+                etallq - et,
+            )
+
+        et_allm = time.time()
+        print(
+            "Time needed for quadratic estimation on all ranks for all m-chunks ",
+            et_allm - et,
+        )
+        # Once done with all the m's, redistribute qa array over m's
+        self.qa = self.qa.redistribute(axis=0)
+
+        # Make an array for local fisher an bias
+        fisher_loc = np.zeros((self.nbands, self.nbands), dtype=np.float64)
+        bias_loc = np.zeros((self.nbands,), dtype=np.float64)
+
+        # Calculate fisher for each m
+        for ml, mg in self.qa.enumerate(axis=0):
+            fisher_m = np.cov(self.qa[ml])
+            bias_m = np.mean(self.qa[ml], axis=1)
+            # Sum over all local m-modes to get the overall fisher and bias per parallel process
+            fisher_loc += fisher_m.real  # be careful with the real
+            bias_loc += bias_m.real
+
+        self.fisher = mpiutil.allreduce(fisher_loc, op=MPI.SUM)
+        self.bias = mpiutil.allreduce(bias_loc, op=MPI.SUM)
+
+        if mpiutil.rank0:
+            et = time.time()
+            print("======== Ending PS calculation (time=%f) ========" % (et - st))
+
+        self.write_fisher_file()
+
+    def q_estimator(self, mi, vec1, vec2, noise=False):
+        """Calculate the quadratic estimator for this mi with data vec1 and vec2"""
+        # if data vector is filled with zeros, return q = 0.0 for this m
+        if np.all(vec1 == 0) and np.all(vec2 == 0):
+            self.qa[mi, :] = 0.0
+
+        # if data vector is empty, return q = 0.0 for this m
+        elif (vec1.shape[0], vec2.shape[0]) == (0, 0):
+            self.qa[mi, :] = 0.0
+
+        else:
+            lside = self.telescope.lmax + 1
+            for bi, bg in self.qa.enumerate(axis=1):
+                for li in range(lside):
+                    lxvec = vec1[:, li]
+                    lyvec = vec2[:, li]
+                    self.qa[mi, bi] += np.sum(
+                        lyvec.conj()
+                        * np.matmul(self.clarray[bi][li].astype(np.complex128), lxvec),
+                        axis=0,
+                    ).astype(np.float64)
+
+        # Calculate q_a for noise power (x0^H N x0 = |x0|^2)
+        if noise:
+            # If calculating crosspower don't include instrumental noise
+            noisemodes = 0.0 if self.crosspower else 1.0
+            evals, evecs = self.kltrans.modes_m(mi)
+            noisemodes = noisemodes + (evals if self.zero_mean else 0.0)
+
+            self.qa.global_slice[mi, -1] = np.sum(
+                (vec1 * vec2.conj()).T.real * noisemodes, axis=-1
+            )
+
+    def recv_send_data(self, data, axis):
+
```

Comment:
Docstring would be useful here, to describe in words what this function does

Reply:
Done.

Reply:
Would need feedback if its comprehensible when I push changes.

Reply:
Looks good!